modules/core/src/test/java/org/apache/ignite/internal/processors/security/sandbox/AbstractContinuousQuerySandboxTest.java

```diff
@@ -0,0 +1,132 @@
+ * Ignite Sandbox.
+ */
+public class AbstractContinuousQuerySandboxTest extends AbstractSandboxTest {
+    /** Cache name for tests. */
+    protected static final String CACHE_NAME = "TEST_CACHE";
+
+    /** Cache index. */
+    private static final AtomicInteger CACHE_INDEX = new AtomicInteger();
+
+    /** Error. */
+    private static volatile AccessControlException error;
+
+    /** Runs control action for CQ tests. */
+    protected static final Runnable CONTROL_ACTION_RUNNER = new Runnable() {
+        @Override public void run() {
+            try {
+                controlAction();
+            }
+            catch (AccessControlException e) {
+                error = e;
+            }
+        }
+    };
+
+    /** Preidacte for inital query tests. */
+    protected static final IgniteBiPredicate<Integer, Integer> INIT_QRY_FILTER = (k, v) -> {
+        CONTROL_ACTION_RUNNER.run();
+
+        return true;
+    };
+
+    /** Remote filter. */
+    protected static final CacheEntryEventSerializableFilter<Integer, Integer> RMT_FILTER = ent -> {
+        CONTROL_ACTION_RUNNER.run();
+
+        return true;
+    };
+
+    /** */
+    protected void checkContinuousQuery(Supplier<Query<Cache.Entry<Integer, Integer>>> s, boolean init) {
+        runOperation(operation(CLNT_ALLOWED_WRITE_PROP, s, init));
+        runForbiddenOperation(operation(CLNT_FORBIDDEN_WRITE_PROP, s, init), AccessControlException.class);
+    }
+
+    /** */
+    private GridTestUtils.RunnableX operation(String nodeName,
+        Supplier<Query<Cache.Entry<Integer, Integer>>> s, boolean init) {
+        return () -> {
+            error = null;
+
+            executeQuery(grid(nodeName), s.get(), init);
+
+            if (error != null)
+                throw error;
+        };
+    }
+
+    /**
+     * Opens query cursor.
+     *
+     * @param ignite Node.
+     * @param q {@link Query}.
+     * @param init True if needing put data to a cache before openning a cursor.
+     */
+    private void executeQuery(Ignite ignite, Query<Cache.Entry<Integer, Integer>> q, boolean init) {
+        IgniteCache<Integer, Integer> cache = ignite.getOrCreateCache(
+            new CacheConfiguration<Integer, Integer>()
+                .setName(CACHE_NAME + CACHE_INDEX.incrementAndGet())
+                .setCacheMode(CacheMode.PARTITIONED)
+        );
+
+        if (init)
+            cache.put(primaryKey(grid(SRV), cache.getName()), 100);
+
+        try (QueryCursor<Cache.Entry<Integer, Integer>> cur = cache.query(q)) {
+            if (!init)
+                cache.put(primaryKey(grid(SRV), cache.getName()), 100);
+
+            cur.getAll();
+        }
+    }
+
+    /**
+     * Getting the key that is contained on primary partition on passed node for cache.
+     *
+     * @param ignite Node.
+     * @param cacheName Cache name.
+     * @return Key.
+     */
+    private Integer primaryKey(IgniteEx ignite, String cacheName) {
```

Comment:
Why can't we use `GridCommonAbstractTest#primaryKey`?

Reply:
I need to get a primary key for a specific node.
GridCommonAbstractTest#primaryKey isn't an appropriate method because it finds a primary key for a local node.

Reply:
`primaryKey(ignite.cache(cacheName))` will solve the problem

Reply:
ok